### PR TITLE
Fixed #29810 -- Fixed crash of select_related() on FilteredRelation with empty result.

### PR DESCRIPTION
**Pull request by cansarigol**
Ticket: [#29810](https://code.djangoproject.com/ticket/29810)

### CODE REVIEW

**Comment by timgraham** [comment]
Can you add a code comment to explain the new code?

**Comment by cansarigol** [comment]
added

**Comment by timgraham** [comment]
The question the comment should answer is "In what case is from_obj None, and why is self.klass_info['related_klass_infos'] the right object to use?`

**Comment by cansarigol** [comment]
@timgraham, is this appropriate?
# In FilteredRelation with condition returns no value cases, form_obj is none and set_cached_value raises 'NoneType' object has no attribute '_state' error. 
# form_obj has to derived from related class of join as a key of caching.

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -894,6 +894,12 @@ def get_related_klass_infos(klass_info, related_klass_infos):
                     from_parent = issubclass(model, opts.model) and model is not opts.model
 
                     def local_setter(obj, from_obj):
+                        if from_obj is None:
+                            # In FilteredRelation with condition returns no value cases, form_obj is none
+                            # and FieldCacheMixin.set_cached_value raises an error about NoneType object.
+                            # form_obj has to be derived from a related class of the join as a key of caching.
+                            if self.klass_info['related_klass_infos']:
+                                from_obj = self.klass_info['related_klass_infos'][0]['model']()
```
Comment (felixxm) [inline django/db/models/sql/compiler.py]: Creating a new empty instance of related `model` doesn't look like a proper solution for me. IMO we should just omit `set_cached_value()` when `from_obj` is not available e.g.
```python
def local_setter(obj, from_obj):
    if from_obj:
        f.remote_field.set_cached_value(from_obj, obj)
```

File: tests/filtered_relation/tests.py [hunk 2]
```diff
@@ -50,6 +50,13 @@ def test_select_related(self):
                 (self.author2, self.book3, self.editor_b, self.author2),
             ], lambda x: (x, x.book_join, x.book_join.editor, x.book_join.author))
 
+    def test_select_related_with_condition(self):
+        qs = Author.objects.annotate(
+            book_join=FilteredRelation('book', condition=Q(book__title__startswith='Poem')),
+        ).select_related('book_join').order_by('pk', 'book_join')
+        with self.assertNumQueries(1):
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: I don't see any value in testing the number of queries in this test.

File: tests/filtered_relation/tests.py [hunk 3]
```diff
@@ -50,6 +50,13 @@ def test_select_related(self):
                 (self.author2, self.book3, self.editor_b, self.author2),
             ], lambda x: (x, x.book_join, x.book_join.editor, x.book_join.author))
 
+    def test_select_related_with_condition(self):
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: `test_select_related_with_condition` -> `test_select_related_with_empty_relation`.

File: tests/filtered_relation/tests.py [hunk 4]
```diff
@@ -50,6 +50,13 @@ def test_select_related(self):
...
+        qs = Author.objects.annotate(
+            book_join=FilteredRelation('book', condition=Q(book__title__startswith='Poem')),
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: I would simplify condition e.g. `condition=Q(pk=-1)`.

File: tests/filtered_relation/tests.py [hunk 5]
```diff
@@ -50,6 +50,13 @@ def test_select_related(self):
...
+        ).select_related('book_join').order_by('pk', 'book_join')
```
Comment (felixxm) [inline tests/filtered_relation/tests.py]: We can remove `'book_join'` from `order_by()`.

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (cansarigol) [reply django/db/models/sql/compiler.py]: in this case, `from_obj` won't be set with None value. this is a problem, isn't it?

**Comment by cansarigol** [comment]
hi, @felixxm I applied your comments.

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (felixxm) [reply django/db/models/sql/compiler.py]: This is reverse relation so your solution will set `None` in a new empty and not really existing instance. `set_cached_value()` should set object for a reverse relation so there is not reason to set it for a mocked empty instance which doesn't exist outside this method. For example, you would basically do this in added test:
```python
Book()._state.fields_cache['author'] = <Author: Alice>
Book()._state.fields_cache['author'] = <Author: Jane>
```

Comment (cansarigol) [reply django/db/models/sql/compiler.py]: Thanks for the explation.

**Comment by heaviss** [comment]
@timgraham Sorry for newbie question, but is there any chance that this fix will be applied to 2.2.x version?

**Comment by timgraham** [comment]
Per our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions), 2.2.x is only receiving fixes for security and data loss issues.

**Pull request merged**